librz/bin/pdb/tpi.c

```diff
@@ -951,1434 +561,1250 @@ static void free_snumeric(SNumeric *numeric) {
+		rz_buf_read_le32(buf, &s->header.TypeIndexEnd) &&
+		rz_buf_read_le32(buf, &s->header.TypeRecordBytes) &&
 
-	stream_file_read(stream, sizeof(STPIHeader), (char *)&tpi_stream->header);
+		rz_buf_read_le16(buf, &s->header.HashStreamIndex) &&
+		rz_buf_read_le16(buf, &s->header.HashAuxStreamIndex) &&
+		rz_buf_read_le32(buf, &s->header.HashKeySize) &&
+		rz_buf_read_le32(buf, &s->header.NumHashBuckets) &&
 
-	base_idx = tpi_stream->header.idx_begin;
+		rz_buf_read_le32(buf, (ut32 *)&s->header.HashValueBufferOffset) &&
+		rz_buf_read_le32(buf, &s->header.HashValueBufferLength) &&
 
-	for (i = tpi_stream->header.idx_begin; i < tpi_stream->header.idx_end; i++) {
-		type = (SType *)malloc(sizeof(SType));
+		rz_buf_read_le32(buf, (ut32 *)&s->header.IndexOffsetBufferOffset) &&
+		rz_buf_read_le32(buf, &s->header.IndexOffsetBufferLength) &&
+
+		rz_buf_read_le32(buf, (ut32 *)&s->header.HashAdjBufferOffset) &&
+		rz_buf_read_le32(buf, &s->header.HashAdjBufferLength);
+}
+
+RZ_IPI bool parse_tpi_stream(RzPdb *pdb, MsfStream *stream) {
+	if (!pdb || !stream) {
+		return false;
+	}
+	if (stream->stream_idx != PDB_STREAM_TPI) {
+		RZ_LOG_ERROR("Error TPI stream index.\n");
+		return false;
+	}
+	pdb->s_tpi = RZ_NEW0(TpiStream);
+	TpiStream *s = pdb->s_tpi;
+	if (!s) {
+		RZ_LOG_ERROR("Error allocating memory.\n");
+		return false;
+	}
+	s->types = NULL;
+	RzBuffer *buf = stream->stream_data;
+	if (!parse_tpi_stream_header(s, buf)) {
+		return false;
+	}
+	if (s->header.HeaderSize != sizeof(TpiStreamHeader)) {
+		RZ_LOG_ERROR("Corrupted TPI stream.\n");
+		return false;
+	}
+	TpiType *type;
+	for (ut32 i = s->header.TypeIndexBegin; i < s->header.TypeIndexEnd; i++) {
+		type = RZ_NEW0(TpiType);
 		if (!type) {
-			return 0;
+			continue;
 		}
-		type->tpi_idx = i;
-		type->type_data.type_info = 0;
-		type->type_data.leaf_type = eLF_MAX;
-		init_stype_info(&type->type_data);
-		if (!parse_tpi_stypes(stream, type)) {
+		type->type_index = i;
+		if (!parse_tpi_types(buf, type) || !type->type_data) {
+			RZ_LOG_ERROR("Parse TPI type error. idx in stream: 0x%" PFMT32x "\n", i);
 			RZ_FREE(type);
+			exit(0);
 		}
-		rz_list_append(tpi_stream->types, type);
+		rz_rbtree_insert(&s->types, &type->type_index, &type->rb, tpi_type_node_cmp, NULL);
 	}
-	return 1;
+	return true;
 }
 
 /**
- * \brief Get SType that matches tpi stream index
- *
+ * \brief Get TpiType that matches tpi stream index
+ * \param stream TPI Stream
  * \param index TPI Stream Index
  */
-RZ_API SType *rz_bin_pdb_stype_by_index(ut32 index) {
+RZ_API RZ_BORROW TpiType *rz_bin_pdb_get_type_by_index(RZ_NONNULL TpiStream *stream, ut32 index) {
+	rz_return_val_if_fail(stream, NULL);
 	if (index == 0) {
 		return NULL;
 	}
 
-	if (is_simple_type(index)) {
-		STypeInfo base_type = parse_simple_type(index);
-		SType *base_ret_type = RZ_NEW0(SType);
-		if (!base_ret_type) {
+	ut64 idx = index;
+	RBNode *node = rz_rbtree_find(stream->types, &idx, tpi_type_node_cmp, NULL);
```

Comment:
`ut32` is now expected by `tpi_type_node_cmp`:
```suggestion
	RBNode *node = rz_rbtree_find(stream->types, &index, tpi_type_node_cmp, NULL);
```